core/domain/translation_domain.py

```diff
@@ -18,9 +18,287 @@
 
 from __future__ import annotations
 
+import copy
+import enum
+
 from core import utils
 
-from typing import Dict
+from typing import Dict, List, Union
+from typing_extensions import TypedDict
+
+
+class TranslatableContentPossibleDataTypes(enum.Enum):
+    """Represents all possible data types for any translatable content."""
+
+    TRANSLATABLE_CONTENT_FORMAT_HTML = 'html'
+    TRANSLATABLE_CONTENT_FORMAT_UNICODE_STRING = 'unicode'
+    TRANSLATABLE_CONTENT_FORMAT_SET_OF_NORMALIZED_STRING = (
+        'set_of_normalized_string')
+    TRANSLATABLE_CONTENT_FORMAT_SET_OF_UNICODE_STRING = 'set_of_unicode_string'
+
+
+class EntityTypesSupportingNewTranslationArch(enum.Enum):
+    """Represents all possible entity types which support new translations
+    architecture.
+    """
+
+    ENTITY_TYPE_EXPLORATION = 'exploration'
+    ENTITY_TYPE_QUESTION = 'question'
+
+
+# The data type for the content in the TranslatableContent/TranslatedContent
+# object.
+ContentInTranslatableContent = Union[str, List[str]]
+
+
+class TranslatableContentDict(TypedDict):
+    """Dictionary representing TranslatableContent object."""
+
+    content_id: str
+    content: ContentInTranslatableContent
+    content_type: TranslatableContentPossibleDataTypes
+
+
+class TranslatedContentDict(TypedDict):
+    """Dictionary representing TranslatedContent object."""
+
+    content: ContentInTranslatableContent
+    needs_update: bool
+
+
+class BaseTranslatableObject:
+    """Base class for all translatable objects which contain translatable
+    fields/objects. For example, a State is a translatable object in
+    Exploration, a Hint is a translatable object in State, and a hint_content
+    is a translatable field in Hint. So Exploration, State, and Hint all should
+    be the child class of BaseTranslatableObject.
+    """
+
+    # Private field to track translatable contents in a BaseTranslatableObject.
+    _translatable_contents: Dict[str, TranslatableContent] = {}
+
+    def _register_all_translatable_fields(self) -> None:
+        """Method to register all translatable fields in a translatable object.
+
+        Raises:
+            NotImplementedError. The derived child class must implement the
+                necessary logic to register all translatable fields in a
+                translatable object.
+        """
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    def _register_translatable_field(
+        self,
+        field_type: TranslatableContentPossibleDataTypes,
```

Comment:
@vojtechjelinek I have a question,
Here I am concerned that if the field_type is an ENUM then we have to access the field by **field_type.value**. So wanted to know whether in real cases we will receive string or ENUM?

Reply:
It depends on how you code it. You can pass both.

Reply:
Got it, thank you.